### PR TITLE
Add drop in xds

### PR DESCRIPTION
**Pull request by AspirinSJL**
Built on https://github.com/grpc/grpc/pull/19394.

Only the last commit is relevant.

---
This change is [Reviewable](https://reviewable.io/reviews/grpc/grpc/19729)

### CODE REVIEW

File: test/cpp/end2end/xds_end2end_test.cc [hunk 1]
```diff
@@ -326,6 +341,18 @@ class EdsServiceImpl : public EdsService {
...
+    if (!drop_categories.empty()) {
+      auto* policy = assignment.mutable_policy();
+      for (const auto& p : drop_categories) {
+        const grpc::string& name = p.first;
+        const uint32_t parts_per_million = p.second;
+        auto* drop_overload = policy->add_drop_overloads();
+        drop_overload->set_category(name);
+        auto* drop_percentage = drop_overload->mutable_drop_percentage();
+        drop_percentage->set_numerator(parts_per_million);
+        drop_percentage->set_denominator(FractionalPercent::MILLION);
```
Comment (markdroth) [inline test/cpp/end2end/xds_end2end_test.cc]: We should also test the code for the other denominators.

File: test/cpp/end2end/xds_end2end_test.cc [hunk 2]
```diff
@@ -1028,6 +1050,176 @@ TEST_F(SingleBalancerTest, AllServersUnreachableFailFast) {
...
+  EXPECT_EQ(1U, balancers_[0]->eds_service()->request_count());
+  // and sent two responses
+  EXPECT_EQ(2U, balancers_[0]->eds_service()->response_count());
+}
+
+TEST_F(SingleBalancerTest, DropAll) {
+  SetNextResolution({}, kDefaultServiceConfig_.c_str());
+  SetNextResolutionForLbChannelAllBalancers();
+  const size_t kNumRpcs = 1000;
+  const uint32_t drop_per_million_for_lb = 100000;
```
Comment (markdroth) [inline test/cpp/end2end/xds_end2end_test.cc]: Let's have this set up such that neither category individually drops 100% but the two categories between them drop 100% (e.g., "lb" drops 20% and "throttle" drops 80%).  That way, we know we'll handle that case right.

File: test/cpp/end2end/xds_end2end_test.cc [hunk 3]
```diff
@@ -1652,7 +1835,71 @@ TEST_F(SingleBalancerWithClientLoadReportingTest, BalancerRestart) {
...
+  const size_t total_rpc = num_warmup + kNumRpcs;
+  EXPECT_THAT(
+      client_stats->dropped_requests("lb"),
+      ::testing::AllOf(
+          ::testing::Ge(total_rpc * drop_rate_for_lb * (1 - error_tolerance)),
+          ::testing::Le(total_rpc * drop_rate_for_lb * (1 + error_tolerance))));
+  EXPECT_THAT(
+      client_stats->dropped_requests("throttle"),
+      ::testing::AllOf(
+          ::testing::Ge(total_rpc * (1 - drop_rate_for_lb) *
```
Comment (markdroth) [inline test/cpp/end2end/xds_end2end_test.cc]: I don't think we should need to account for `drop_rate_for_lb` when looking at `drop_rate_for_throttle`.  See my comment elsewhere about not wanting to do a different roll of the dice for each drop category.

**Comment by AspirinSJL** [comment]
Fixed a bug so that if the drop config changes while the locality list doesn't we still update the xds picker to use the new drop config.

**Comment by AspirinSJL** [comment]
Basic Tests C/C++ MacOS: https://github.com/grpc/grpc/issues/19802, https://github.com/grpc/grpc/issues/19819
Basic Tests Node MacOS: https://github.com/grpc/grpc/issues/19422
Interop Cloud-to-Cloud Tests (internal CI): https://github.com/grpc/grpc/issues/13379

**Pull request merged**